File: app/src/main/java/se/ntlv/newsbringer/NewsThreadListAdapter.java
package se.ntlv.newsbringer;

import android.content.Context;
import android.database.Cursor;
import android.view.View;
import android.widget.ResourceCursorAdapter;
import android.widget.TextView;

import se.ntlv.newsbringer.database.PostTable;


class NewsThreadListAdapter extends ResourceCursorAdapter {

	public NewsThreadListAdapter(Context context, int layout, Cursor c, int flags) {
		super(context, layout, c, flags);
	}

	static String get(Cursor cursor, String columnName) {
		return cursor.getString(cursor.getColumnIndexOrThrow(columnName));
	}

	static ViewHolder getOrMakeTag(View view) {
		ViewHolder tag = (ViewHolder) view.getTag();
		if (tag == null) {
			ViewHolder newTag = new ViewHolder(view);
			view.setTag(newTag);
			return newTag;
		} else {
			return tag;
		}
	}

	@Override
	public void bindView(View view, Context context, Cursor cursor) {
		ViewHolder tag = getOrMakeTag(view);

		tag.title.setText(get(cursor, PostTable.COLUMN_TITLE));
		tag.by.setText(get(cursor, PostTable.COLUMN_BY));
		tag.ordinal.setText(get(cursor, PostTable.COLUMN_ORDINAL));
		tag.time.setText(get(cursor, PostTable.COLUMN_TIMESTAMP));
		tag.score.setText(get(cursor, PostTable.COLUMN_SCORE));
		tag.link = get(cursor, PostTable.COLUMN_URL);
	}

	static class ViewHolder {

		public TextView title;
		public TextView by;
		public TextView ordinal;
		public TextView time;
		public TextView score;
		public String link;

		public ViewHolder(View root) {
			title = (TextView) root.findViewById(R.id.title);
			by = (TextView) root.findViewById(R.id.by);
			ordinal = (TextView) root.findViewById(R.id.ordinal);
			time = (TextView) root.findViewById(R.id.time);
			score = (TextView) root.findViewById(R.id.score);
		}
	}
}

File: app/src/main/kotlin/se/ntlv/newsbringer/NewsThreadListAdapter.kt
package se.ntlv.newsbringer

import android.database.Cursor
import android.content.Context
import android.widget.ResourceCursorAdapter
import android.view.View
import android.widget.TextView
import se.ntlv.newsbringer.database.PostTable
import se.ntlv.newsbringer.NewsThreadListAdapter.ViewHolder
import se.ntlv.newsbringer.network.NewsThread.Metadata

open class NewsThreadListAdapter(ctx: Context, layout: Int, cursor: Cursor?, flags: Int) :
        ResourceCursorAdapter(ctx, layout, cursor, flags) {

    override fun bindView(view: View, context: Context?, cursor: Cursor) {
        val tag = view.tag
        tag.title.setText(cursor.getString(PostTable.COLUMN_TITLE))
        tag.by.setText(cursor.getString(PostTable.COLUMN_BY))
        tag.time.setText(cursor.getString(PostTable.COLUMN_TIMESTAMP))
        tag.score.setText(cursor.getString(PostTable.COLUMN_SCORE))
        tag.link = cursor.getString(PostTable.COLUMN_URL)
        tag.id = cursor.getLong(PostTable.COLUMN_ID)
        tag.text = cursor.getString(PostTable.COLUMN_TEXT)
        val children = cursor.getString(PostTable.COLUMN_CHILDREN)
        if (children.length() > 0) {
            val strings = children.split(',')
            tag.commentCount.setText(strings.size().toString())
            tag.commentQuantity = strings.size().toLong()
        } else {
            tag.commentCount.setText("0")
            tag.commentQuantity = 0L
        }
    }

    fun Cursor.getString(columnName: String): String = getString(getColumnIndexOrThrow(columnName))
    fun Cursor.getLong(columnName: String): Long = getLong(getColumnIndexOrThrow(columnName))


    class ViewHolder(root: View) {
        val title = root.findViewById(R.id.title) as TextView
        val by = root.findViewById(R.id.by) as TextView
        val time = root.findViewById(R.id.time) as DateView
        val score = root.findViewById(R.id.score) as TextView
        var link: String? = null
        var id: Long? = null
        var text: String? = null
        val commentCount = root.findViewById(R.id.comment_count) as TextView
        var commentQuantity: Long? = null

        fun TextView.getContent(): String = this.getText().toString()

        val metadata: Metadata
            get() = Metadata(id as Long, text ?: "", title.getContent(), by.getContent(), time.getContent(), score.getContent(), link, commentQuantity ?: 0)
    }

    val View.tag: ViewHolder
        get() {
            val tag = getTag()
            if (tag == null) {
                val newTag = ViewHolder(this)
                this.setTag(newTag)
                return newTag
            } else {
                return tag as ViewHolder
            }
        }
}


